### ISSUE TITLE
How To Compile

### ISSUE THREAD

**Issue by MarioTainaka** [UltimateNova1203/mgs2-pchd#2]
Forgive me for asking but does a release build of this actually exist? if not which Powershell script do I need to run to start the compilation process?

**Comment by UltimateNova1203** [comment]
Forgive me for leaving this in a pretty unfinished state, the installers don't do much at the moment outside of backing up the files that will be edited.

You will need to use Solidus to extract the XTI files from the 'cache.qar' files. IIRC, it essentially just deflates the container. You should be able to open the 'cache.qar' file, select all the files, and be able to extract the XTI files, leaving them in the XTI format.

From there, you can edit the CNF files for each stage, replacing the "@cache.qar" line with an entry for each XTI texture. The game will load them and isn't too picky for this. Convenient for what we need to do.

For example, the 'data.cnf' file located in the directory 'cdrom.img/stage/n_title' contains the below code:

`.nocache
.cache
@cache.qar
cache.dar
scenerio.gcx`

The line '@cache.qar' contains all the textures used for that stage. Each stage is any point before a loading screen, or fade to a black screen as it loads the next bit.

So if we replace the '@cache.qar' with the XTI files, we would get something like this. It isn't the full list for the 'data.cnf' replacement, but just a short snippet.

`.nocache
.cache
@a_b_itl_0_dummy.xti
@a_b_spn_0_dummy.xti
@aaa_sp_menu_0_full.xti
@adjust_tex_0_dummy.xti
@boss_opt_0_full.xti
@boss_rush2_0_full.xti
@boss_rush_0_full.xti
@casting_0_full.xti
@color_0_full.xti
cache.dar
scenerio.gcx`

Now that the 'data.cnf' file is prepped, you can load the game, and nothing should have changed. If you use something like ProcMon, you should be able to see it loading all the individual XTI files, rather than just once with the 'cache.qar' file.

Now that the files are prepped in the right way, we can actually replace the texture data. You need to use Solidus to replace the XTI texture's data with the supplied PNG. It should be as simple as just telling it to replace the texture in the tool, but there's nuance in the game engine that prevents us from just doing that, which I'll explain a little further down.

Once the texture in the XTI file has been swapped out, you'll have the original XTI file backed up with a .BAK extension, and the newer much larger file with the same name, without the .BAK extension. If properly done, we can just load the game and it'll use the newer texture.

--------------

Here's where issues start to occur. Since this is a game that was converted from the original Xbox release, it's stuck doing things the way the Xbox did. The game loads files as it needs them. We can replace any textures as we want, but we will quickly run into an arbitrary limit as the game wasn't really designed with multiple texture quality in mind.

It seems we have about 16MB of total space for asset's to load for each stage. This is all models, textures, audio, scenario files, everything. So if we replace all the textures for a stage with double scaled textures (like what was done in the HD Editions), we will quickly exceed 16MB from textures alone. The game will load fine, but the second that 16MB is exceeded on loading assets, it'll soft lock.

An example is on the Strut A Roof during the beginning of the plant. If we double scale half of the used assets in that stage, then when we get the codec call from the Colonel regarding the Seals attack, the codec screen loads, but the call never starts, since we exceeded our max size.

So now we are stuck with only some textures being able to be upscaled so we don't exceed our budget. The only way to get around this is to have the game's code patched. I'd imagine someone like nemesis2000 or vFansss could find some way to do this, by adding some virtual memory pointers (or something like that, I'm not a knowledgable coder), to redirect calls to the texture to another address space. So our normal budget won't be exceeded, but we can use whatever arbitrary amount of textures we want.

Currently, a workaround is to truncate the XTI files. So if we have a texture that's 1024x1024, but it only uses the first 300px of the textures height, we can truncate the PNG files to see how much space we can shave off, then truncate the full size XTI file to reclaim some space. It's a bit tedious work, and having the Solidus source code would immensely help with optimizing the XTI textures (or just someone to fully document the XTI format).

This is partially why the project stalled for so long, and I lost interest (and PC crash got rid of my notes on XTI format) for the time being. If someone can do some bluepoint magic, or pull a page from the modding community for GTAIII/VC/SA for their texture replacement, I'd be more than willing to get this into an actual usable state.

Tl;Dr, you'll have to manually compile the textures with Solidus, and point the game to the new textures. Unless we get a replacement for Solidus, there is a lot of manual intervention at the moment.

**Comment by MarioTainaka** [comment]
thank you! I didn't see this till now. What should I do if I just feeling like wanting the new HD fonts?

**Comment by UltimateNova1203** [comment]
You'll want to use the Solidus tool to unpack the CACHE.QAR files to get the raw XTI textures. Edit the data.cnf file for the stage in question (n_title is the main title screen on a return from the VR Missions or Snake Tales menu's), and manually replace the textures. I'm looking into possibilities on what I can do without directly modifying game code, but it's fairly limited on what is possible.

**Comment by MarioTainaka** [comment]
Yikes... that's alot to take in. Knowing my luck i'll probably break something
